Adds block about Angular's low-level functions

sergiocruz opened the pull request:
Angular 1.x low-level functions should be avoided in favor of language features that are available natively in its newest versions (ES2015+). For most other cases a utility library like lodash will do wonders as well.

Although it was not mentioned in the guide, avoiding low-level also helps when upgrading to Angular 2. With the help of TypeScript and newest language features, Angular 2 is able to shift more responsibility to the language itself. Therefore functions like `angular.copy()`, `angular.isUndefined()` and `angular.equals()` is simply not found within Angular 2.

ericcarraway committed:
fix typo: declaration => declarations

anteriovieira committed:
Original text and redundant removed

andreyshatilov committed:
Fixed typo

sergiocruz committed:
adds block about angular low-level functions

Update README.md

Update README.md

Addressed comments made in johnpapa/angular-styleguide#768

Low level functions: added code samples and more why's

Angular Helper Functions: more reviews in johnpapa/angular-styleguide#768

Angular Helper Functions: addressing more comments on johnpapa/angular-styleguide#768

Helper functions: Changed language to sound more like a soft suggestion

joshstar commented:
I feel it would be better if this was prefaced "If you plan on migrating to Angular 2 or using ES2015+". There are a lot of angularjs apps that follow this guide that have but have no plans on moving up to Angular 2, and browser support limits a lot of them from using ES2015 unless a polyfill or something like Typescript is used.

gkalpak commented on a1/README.md:
```diff
@@ -3268,6 +3269,16 @@ Use [Gulp](http://gulpjs.com) or [Grunt](http://gruntjs.com) for creating automa
+   
+  *Why?*: You can replace a lot Angular's low-level functions with native ES2015+ functionality such as [`Object.assign()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) for copying objects and the [Spread Syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator#Copy_an_array) for copying arrays _(ie. `[. ...
+
+  - Note: Use a utility library like [lodash](https://lodash.com/docs/4.15.0) if you need additional functionality that is not available natively in the newest versions of JavaScript.
```
It is better to omit `4.15.0` from the link (to always point to the latest version).

johnpapa commented:
Please add a style code for this too. See other examples. I think where you put this is fine, i dont see a section that fits it, but i do like this new topic.

Perhaps style Y500

I'll read through it when I get a chance ... til then, love to get other people's feedback too.

BTW - we still need a style for components :)

sergiocruz commented:
I addressed some of the comments made here, and pushed a new (squashed commit) adding:

- intro paragraph
- code samples
- style number
- some other revisions

Let me know what you all think :)

johnpapa commented on a1/README.md:
```diff
@@ -3268,6 +3269,34 @@ Use [Gulp](http://gulpjs.com) or [Grunt](http://gruntjs.com) for creating automa
+
+Angular 1.x ships with some [helper functions](https://docs.angularjs.org/api/ng/function) that are found within JavaScript natively. In such cases we should favor usage of native JavaScript functions instead of Angular's helper functions. In other cases where functionality is not found in JavaScript natively, we should turn to utility libraries such as [lodash](https://lodash.com/docs) to fill-i ...
+
+|Replace | With | |
```
Great examples ... net step is to put these examples as code examples, and in the format of the avoid/recommend that the guide uses.

sergiocruz replied:
No problem! I decided to use a table here as it was a lots os micro-examples, but will move to that style then (will probably trim down the examples though).

johnpapa commented on a1/README.md:
```diff
@@ -3268,6 +3269,34 @@ Use [Gulp](http://gulpjs.com) or [Grunt](http://gruntjs.com) for creating automa
+
+  *Why?*: You can replace a lot Angular's low-level functions with native ES2015+ functionality such as [`Object.assign()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) for copying objects and the [Spread Syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator#Copy_an_array) for copying arrays _(ie. `[. ...
+
+  *Why?*: This will ease the process of upgrading to Angular 2 as it does not ship with these helper functions.
```
others why's ... (needs wordsmithing) ...

- Because it's baked into ES. They were written at a time when they were not readily availableas they are today, thus there is no need for a framework specific wrapper. 

- These ES features are already more commonly used, as they are ES native.

johnpapa commented:
coming along well!

sergiocruz commented:
I added another commit adding the code samples and the two additional _why's_.

gkalpak commented on a1/README.md:
```diff
@@ -3268,6 +3269,95 @@ Use [Gulp](http://gulpjs.com) or [Grunt](http://gruntjs.com) for creating automa
 
+## Low Level Functions
+
+Angular 1.x ships with some [helper functions](https://docs.angularjs.org/api/ng/function) that are found within JavaScript natively. In such cases we should favor usage of native JavaScript functions instead of Angular's helper functions. In other cases where functionality is not found in JavaScript natively, we should turn to utility libraries such as [lodash](https://lodash.com/docs) to fill-i ...
```
I think it should be (somehow) clarified that not all. functions listed in the linked docs section are to be avoided. E.g. `module`, `injector`, `bootstrap`, `element` are fine to use.

sergiocruz replied:
That's a very good point. I wonder how to balance that with the fact that if one ever wants to upgrade to Angular 2, he or she will have to get rid of these method calls anyway.

Just mentioning this since eventual upgrade to Angular 2 is one of the _why's_.

gkalpak commented on a1/README.md:
```diff
@@ -3268,6 +3269,95 @@ Use [Gulp](http://gulpjs.com) or [Grunt](http://gruntjs.com) for creating automa
+    console.log(value);
+  });
+
+  // available in ES
```
ES what? :smile:

gkalpak commented on a1/README.md:
```diff
@@ -3268,6 +3269,95 @@ Use [Gulp](http://gulpjs.com) or [Grunt](http://gruntjs.com) for creating automa
+
+  *Why?*: Because this is functionality that has been made available in the latest versions of JavaScript, but did not exist natively back when helper functions were first written.
+
+  Replace `angular.forEach()` and `angular.isArray()`:
```
It might be worth noting that there _are_ slight differences between Angular's and native/lodash methods (especially in less common cases).

E.g. `angular.forEach` also supports array-like objects and won't barf when called with `null`/`undefined`

sergiocruz replied:
I believe the note I added in the bottom of the block also addresses this concern.

gkalpak commented on a1/README.md:
```diff
@@ -3268,6 +3269,95 @@ Use [Gulp](http://gulpjs.com) or [Grunt](http://gruntjs.com) for creating automa
+
+  var obj1 = { hello: 'Hello' };
+  var obj2 = { world: 'World' };
+  var obj3 = angular.extend({}, obj1, obj2); // {hello: 'Hello', world: 'World'}
```
Again, might not be a drop-in replacement, since `angular.extend` does for example preserve the destination's `$$hashKey`, which might result in different behavior in some cases. Not suggesting that such low-level details should be mentioned in the styleguide, just reinforcing the argument for adding a general note that - while you have to be aware of the fact that there might be slight differences - moving away from these helpers is the right thing to do (in many cases).

sergiocruz replied:
I added a note for this later in the block advising people to understand the differences in implementation.

gkalpak commented on a1/README.md:
```diff
@@ -3268,6 +3269,95 @@ Use [Gulp](http://gulpjs.com) or [Grunt](http://gruntjs.com) for creating automa
+
+  - Note: Favor usage of functions that are available in JavaScript natively, then resort to using a utility library like [lodash](https://lodash.com/docs) when what you need is not available.
+
+  - Note: you might need to use a [polyfill](https://babeljs.io/docs/usage/polyfill/) to take advantage the latest features of JavaScript for better browser support.
```
_take advantage **of** the latest_

sergiocruz commented:
All righty, added another commit addressing some more of the comments including:

- Renamed "Low-level Functions" to "Helper Functions" (I believe the meaning is clearer now)
- Added a couple of helper fn examples in the intro paragraph so users know what type of functions we are referring to when we mention helper functions
- Added a note advising users to understand the differences in implementation

Again, let me know what you all think :)

gkalpak commented on a1/README.md:
```diff
@@ -3268,6 +3269,98 @@ Use [Gulp](http://gulpjs.com) or [Grunt](http://gruntjs.com) for creating automa
+
+  - Avoid using Angular helper functions such as `angular.copy()`, `angular.extend()` or even `angular.isUndefined()`.
+
+  *Why?*: You can replace a lot Angular's helper functions with native ES2015+ functionality such as [`Object.assign()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) for copying objects and the [Spread Syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator#Copy_an_array) for copying arrays _(ie. `[...a ...
```
_a lot **of** Angular's_

gkalpak commented on a1/README.md:
````diff
@@ -3268,6 +3269,98 @@ Use [Gulp](http://gulpjs.com) or [Grunt](http://gruntjs.com) for creating automa
+  var obj3 = Object.assign({}, obj1, obj2); // {hello: 'Hello', world: 'World'}
+  ```
+
+  *Why?*: These are commonly known features of JavaScript and makes code more intuitive to read.
````
"and ***make*** code" or "and ***using them*** makes code"

gkalpak commented on a1/README.md:
```diff
@@ -3268,6 +3269,98 @@ Use [Gulp](http://gulpjs.com) or [Grunt](http://gruntjs.com) for creating automa
+
+  - Note: you might need to use a [polyfill](https://babeljs.io/docs/usage/polyfill/) to take advantage of the latest features of JavaScript for better browser support.
+
+  - Note: Be sure to understand the differences in implementation when using a native JavaScript feature over an Angular helper. Some of Angular's helpers have some nuances that are relevant to its internal API (ie. dealing with `$$hashKey`, etc).
```
The example with the (undocumented, private) `$$hashKey` will sound scary to the uninitiated :smile:
I would rather give the example of Angular helpers being more "forgiving" (e.g. not throwing on `undefined`/`null` values).

gkalpak commented:
Another reason you might want to mention is:

These helpers are exposed publicly for convenience, but they are primarily intended for Angular's internal use. What this means is that, while they are optimized for Angular's needs, they may not handle all edge-cases (or handle specific - not interesting to Angular - cases sub-optimally).

sergiocruz commented:
Thank you, I just pushed another commit.

sergiocruz commented:
Hey @johnpapa are there any reviews left to add before we are able to merge this in?

johnpapa commented:
I want to get Pete Bacon Darwin's eyes on this too

gkalpak commented:
/summon @petebacondarwin

petebacondarwin commented:
While I would be delighted if we had never exposed our "helper" functions to application in the first place, because it would make our maintenance of the project easier, I am still not clear how it benefits application developers not to use them.

Sure, if you are going to upgrade to Angular 2 then you would need to change to alternative helpers. But compared the the significant changes that will need to be made across the board for an app migration, such as completely reworking any directives that directly manipulate the DOM, I think that the cost of swapping out things like `angular.isArray` with `Array.isArray` is pretty trivial.

Also, I am not entirely convinced that there is much of a performance saving in many cases either. Often the helper function delegates to the native method if it is available anyway.

Moreover, as is pointed out by @gkalpak, the helpers often work differently to their native counterparts, which often leads to simpler application code. E.g.

```
angular.forEach(obj, function(value) {
  console.log(value);
});
```

compared to 

```
Object.keys(obj).forEach(function(key) {
  console.log(obj[key]);
});
```

The only reason that I feel holds water is that using standard native JavaScript functions rather than Angular helpers allows more idiomatic JavaScript code that is easier for a non-Angular developers to follow.

So I guess my opinion is that I am not against people using native methods if they like, but I don't feel strongly that people should always avoid using them. So I have no problem with this being added to the style guide but personally I would make it a very soft suggestion rather than a strong recommendation.

johnpapa commented:
I agree with you Pete ... especially in this
> The only reason that I feel holds water is that using standard native JavaScript functions rather than Angular helpers allows more idiomatic JavaScript code that is easier for a non-Angular developers to follow.

And also with this ... 

...  I would make it a very soft suggestion rather than a strong recommendation.

sergiocruz committed:
Merge pull request #3 from johnpapa/master

Getting the latest

sergiocruz commented:
@johnpapa I pushed a new commit changing the language a bit to make it sound more like a soft suggestion with examples. Hopefully this is better, but still open to additional suggestions :)

petebacondarwin commented:
Reads better for me.

johnpapa commented:
Great insight about avoiding Angular's low-level functions in favor of native ES2015+ features. Since the Angular 1 guide has stabilized and Angular 1 has reached end of life, we're not adding new rules at this point. The advice is solid though — thank you for contributing it! 🙏
